Admin: migrate users management to GraphQL, TSify

nl0 opened the pull request:
# Description


# TODO

<!-- Remove items that are irrelevant to this PR -->

- [ ] Unit tests
- [ ] Automated tests (e.g. Preflight)
- [ ] Confirm that this change meets security best practices and does not violate the security model
- [ ] Documentation
    - [ ] [Python: Run `build.py`](../tree/master/gendocs/build.py) for new docstrings
    - [ ] JavaScript: basic explanation and screenshot of new features
    - [ ] Markdown somewhere in docs/**/*.md that explains the feature to end users (said .md files should be linked from SUMMARY.md so they appear on https://docs.quiltdata.com)
    - [ ] Markdown docs for developers
- [ ] [Changelog](../tree/master/docs/CHANGELOG.md) entry (skip if change is not significant to end users, e.g. docs only)

nl0 commented:
@fiskus i think this is ready for review, pls have a look when you get some time. we still need to update the gql schema once it's finalized on the backend and regenerate all the dependent code, but most likely it won't affect anything in this PR

fiskus reviewed (commented):
I think list divider looks nicer with small margin like in default MUI List view:

![Screenshot from 2024-06-11 13-41-33](https://github.com/quiltdata/quilt/assets/533229/44aaa36c-6f0d-427a-9530-834327a13186)

![Screenshot from 2024-06-11 13-41-13](https://github.com/quiltdata/quilt/assets/533229/cfbe43b7-47ed-496a-a19e-87664d15409d)

---

I didn't fully pay attention to your conversation with Aneesh. This is my feedback. Maybe, you already discussed this.

"Switch Role" dialog is fine and simple: I totally understand how to use. But first impression is it kinda broken. There are radio buttons, but you can't select it and do something with the selection, it acts like a link. I would expect:
* no radio buttons ("check" icon for selected is fine)
* or leave radio buttons and add a submit button

Dialog with submit looks "safer" for me. Because, I have a transition between my actions and commitment to actions. And if something goes wrong, I will see the same form with error message.

Also, small padding at the bottom would be nice (like below "Spam" in the screenshot)
![image](https://github.com/quiltdata/quilt/assets/533229/564a48f4-5a53-46a7-a20c-de3f71368982)

I don't expect to change anything except padding though. Just wanted to share my impression.
---

Hint "(you)" is really cool small helpful thing.

nl0 commented:
@fiskus thanks for the feedback

> I think list divider looks nicer with small margin like in default MUI List view:

well, this is default list view, i didn't tweak the padding (maybe it's because of using `<ListItem divider ... />` vs `<Divider />` 🤷 )

> radio buttons and add a submit button

yeah i was thinking about the explicit confirmation, but then i guess i just forgot about it.
i don't mind adding the button if that's the consensus, tho i don't see the current ux as problematic either.
@akarve wdyt?

> Also, small padding at the bottom would be nic

i'll consider this

> Hint "(you)" is really cool small helpful thing.

👍

nl0 commented:
@fiskus i've added a submit button to role switching dialog, some spacing around menu separators, and incorporated your change from #3704 -- pls have a look.
i'll deploy this to the dev stack shortly.

fiskus reviewed (commented):
UX:

* When I open "Assign roles to " dialog, I'm not sure what a selection with radio button does. I guess, it switches the role. After a couple of minutes, it absolutely makes sense. But I have initial frustration here. Probably, because I didn't *want* to change the role, and just click around :) Maybe, tooltip "Switch role to %role name%" could help.
* "Assigned" vs "Available" confused me a little. Maybe, rename it to "Assigned"/"All". As I understand, "available" -  all roles available for the stack, and "assigned" - available for the user to switch.

Other than that, all looks and feels good

fiskus commented:
I noticed, that "Users" table now looks prettier. Though, I wasn't sure. Then I compared to master: yes, it looks much nicer.
I think, you can also remove "Actions" title. Then "+" and "Delete"  buttons become confront and consistent to each other at the same time. Also, it gives slightly more space to fit into a small viewport.

I would re-use the same dialog for adding and editing role in favor to have fewer code. So, click on row can open edit dialog. But, clicking on individual items makes more sense.

nl0 commented:
> Maybe, tooltip "Switch role to %role name%" could help.

actually there is a tooltip when you hover over a radio button (tho it's easy to miss, i agree)

nl0 commented:
> rename it to "Assigned"/"All".  As I understand, "available" - all roles available for the stack,

well, "available" != "all" ("available" = "all" - "assigned")

> and "assigned" - available for the user to switch.

correct

nl0 commented:
> I noticed, that "Users" table now looks prettier. Though, I wasn't sure. Then I compared to master: yes, it looks much nicer.

yeah, i tried my best to polish it

> I think, you can also remove "Actions" title. Then "+" and "Delete" buttons become confront and consistent to each other at the same time. Also, it gives slightly more space to fit into a small viewport.

good idea, i'll try that

> I would re-use the same dialog for adding and editing role in favor to have fewer code. So, click on row can open edit dialog. But, clicking on individual items makes more sense.

i don't quite get that. you mean "invite" vs "assign roles" dialogs?

fiskus commented:
> i don't quite get that. you mean "invite" vs "assign roles" dialogs?

Yes "invite", sorry.
"Invite user" has "username" + "email" + "assign role" fields, and "Edit user" could have "email" + "assign role" fields.
So, "Invite/Edit user" dialog could be:
```
{isNew ? "Username" : null}
{"Email"}
{"Assign roles"}
```

nl0 commented:
> > I think, you can also remove "Actions" title. Then "+" and "Delete" buttons become confront and consistent to each other at the same time. Also, it gives slightly more space to fit into a small viewport.
> 
> good idea, i'll try that

it looks ok

nl0 commented:
> > i don't quite get that. you mean "invite" vs "assign roles" dialogs?
> 
> Yes "invite", sorry. "Invite user" has "username" + "email" + "assign role" fields, and "Edit user" could have "email" + "assign role" fields. So, "Invite/Edit user" dialog could be:
> 
> ```
> {isNew ? "Username" : null}
> {"Email"}
> {"Assign roles"}
> ```

well, yeah, i thought about unifying all the user editing functionality under a single screen / dialog, but i decided to keep things as they were (focused "atomic" editing interfaces).
maybe next time we revisit this ui we'll get to unifying those dialogs.

fiskus commented on catalog/app/containers/Admin/UsersAndRoles/Users.tsx:
```diff
@@ -0,0 +1,1070 @@
+
+const useDialogFormStyles = M.makeStyles((t) => ({
+  root: {
+    marginTop: t.spacing(-3),
```
I think "(-2)" is nicer. So, I think, the padding for title and dialog content is ok. The only problem is top margin for the first text field, and it's a 16px, not 24px.
Otherwise, the margin under the dialog title ("Invite user") and under the section title ("Assign roles") looks the same. But, dialog title margin seems intuitively should be bigger.

nl0 replied:
i see the logic behind this, but -3 _feels_ better. well, whatever

nl0 commented on catalog/app/containers/Admin/UsersAndRoles/Users.tsx:
```diff
@@ -0,0 +1,1070 @@
+    rows,
+    filterBy: ({ email, name }) => email + name,
+  })
+  const ordering = Table.useOrdering({
```
well, that would interfere with the sorting logic (the rows are ordered by a selected column / direction)

nl0 commented on catalog/app/containers/NavBar/NavBar.tsx:
```diff
@@ -486,176 +177,64 @@ export function Container({ children }: ContainerProps) {
-      ) : (
-        <Controls />
-      )}
+      {hideControls ? <div className={classes.spacer} /> : <Controls />}
```
not sure it will work out ok with the current flexbox layout

nl0 commented on catalog/app/containers/NavBar/NavMenu.tsx:
```diff
@@ -0,0 +1,562 @@
+  collapse: boolean
+}
+
+export function Menu({ collapse }: MenuProps) {
```
yeah i tried hard to achieve some balance between DRYness and comprehensibility
